### DISCUSSION TITLE
Files that don't extract properly

### DISCUSSION THREAD

**Discussion by drsn00ker** [onekey-sec/unblob#421] [General]
I have a few firmware images that don't extract properly with unblob (nor binwalk). Would it be beneficial for the developers to have copies of them to hone the extraction algorithm? If yes, perhaps we can use this discussion to post links to currently unsupported files?

**Comment by kissgyorgy** [comment]
Yes it would be probably helpful, but I can't promise we will implement every format :) If it's a very common format or a bug, we will definitely fix it.

**Reply by drsn00ker** [reply]
I'll add a few more firmwares that do not extract using neither binwalk nor unblob. These are ZTE H298A and H198A
https://pd.trudmore.ru/index.php/s/D29C4NDaxKy8ZmC/download
https://pd.trudmore.ru/index.php/s/j2Ncrjm4XF9JXsB/download

from:
https://manual.trudmore.ru/zte-zxhn-h298a-hw-v1-1-20/
https://manual.trudmore.ru/zte-h198a-v9-hw-9-0-10/

**Comment by drsn00ker** [comment]
I guess I'll start with a few centurylink modems
http://internethelp.centurylink.com/internethelp/modems/pk5000/firmware/QAP002-3.60.3.0.9.9-PK5000.img
http://internethelp.centurylink.com/internethelp/modems/q2000/firmware/QAB001-33.00L.16.img
https://internethelp.centurylink.com/internethelp/modems/q1000/firmware/QAQ004-31.20L.7.img

**Comment by drsn00ker** [comment]
or the zyxel AMG1302-T15C
https://drive.google.com/file/d/1j7BayPwmNZekHgzMzCBhBUNW0VV1E-Ek/view

**Comment by drsn00ker** [comment]
Decided to do a real comparison between binwalk and unblob. First downloaded all the files from ftp://ftp.zyxel.com.tr (that's 470 different firmware files). Ran them against binwalk and unblob. 
Binwalk manages to extract the file system from 10 more files than unblob (see binwalk_only.txt in attachment) 
Unblob and Binwalk extract the file system of 101 FW (see both.txt)
Unblob never manages to extract a file system when binwalk fails as well.
That still leaves 359 firmwares unextractable. (see neither.txt) 
Perhaps those 359 files would be a good start for the developers of unblob to be able to distinguish their program from binwalk!

[extractables.zip](https://github.com/onekey-sec/unblob/files/11782883/extractables.zip)

I'll leave my bash script here as it took me hours to figure out how to deal with spaces and parenthesis in the filenames and still run those through the extractor.


```sh
#!/bin/bash
find . -name '*.zip' > list.txt
find . -name '*.bin' >> list.txt
find . -name '*.rar' >> list.txt
find . -name '*.img' >> list.txt

exec {file_descriptor}<"./list.txt"
declare -a input_array

while IFS=$'\n' read -a input_array -u $file_descriptor
	do
	unblob "$input_array"
	done

exec {file_descriptor}>&-
```
